File: runtime/typescript/broadcasts.ts
// Broadcasts — turbo-stream emit bridge.
//
// The model lowerer's `broadcasts_to` expansion produces calls like
// `Broadcasts.prepend({ stream, target, html })` from inside model
// callback methods (`after_create`, `after_update`, etc.). This shim
// wraps those calls in the turbo-stream wire format and delegates
// to whatever `broadcast(stream, html)` is provided by the active
// `juntos.ts` variant:
//
//   - juntos.ts (better-sqlite3 / sync): null until the HTTP server
//     installs a CableServer-backed broadcaster via `setBroadcaster`.
//     Pre-install calls (and test runs that never install) are
//     silent no-ops, matching the prior shim behavior.
//   - juntos-libsql.ts (libsql / async): same shape — the server
//     installs a CableServer-backed broadcaster.
//   - juntos-worker.ts (SharedWorker): a default
//     BroadcastChannel-backed broadcaster is installed at module
//     load (no setBroadcaster needed); broadcasts reach all tabs
//     of the same origin natively.
//
// One slot, one install path. Previously this file kept its own
// `_broadcaster` slot and exposed `installBroadcastsBroadcaster`
// — neither was wired up anywhere, making `broadcasts_to` a silent
// no-op on every target. Surfaced by the multi-tab smoke probe.

import { broadcast } from "./juntos.js";

interface BroadcastOpts {
  stream: string;
  target?: string;
  html?: string;
  // Rendered attribute text carrying its own leading space —
  // ` maintain_scroll="true"`. Composed by the broadcast lowering, where
  // the literal hash the app wrote can be read; a string rather than a
  // hash so no target needs a markup renderer of its own.
  attributes?: string;
}

function emit(action: string, opts: BroadcastOpts): void {
  const target = opts.target ?? opts.stream;
  const html = opts.html ?? "";
  // Match the turbo-stream wire format the cable speaks: a tagged
  // `<turbo-stream action="…" target="…">…</turbo-stream>` element.
  // `attributes` rides ahead of `action`/`target`, where turbo-rails'
  // `tag.turbo_stream(template, **attributes, action:, target:)` puts it.
  const attributes = opts.attributes ?? "";
  const fragment =
    `<turbo-stream${attributes} action="${action}" target="${target}">` +
    `<template>${html}</template>` +
    `</turbo-stream>`;
  broadcast(opts.stream, fragment);
}

/** Static-method facade matching the lowerer's typing stub
 *  (see `broadcasts_class_info` in `model_to_library`). Each method
 *  takes a kwargs-shaped hash and returns nothing. The action set
 *  matches `BroadcastAct` in the lowerer (Append/Prepend/Replace/
 *  Remove) — adding methods here without a matching variant there
 *  produces dead code. */
export class Broadcasts {
  // Compose the `<turbo-stream>` element for a `turbo_stream.<action>`
  // call in a `.turbo_stream.erb` template. Positional and String-typed on
  // purpose: it is the ONE shape the view lowerer emits on every target
  // (the older keyword/Symbol `render_fragment` spellings had drifted
  // apart between targets).
  //
  // `attributes` carries its own leading space and is written BEFORE
  // `action`/`target`, where turbo-rails' `tag.turbo_stream(template,
  // **attributes, action:, target:)` puts it. Optional, so the
  // three-argument call the view lowerer emits is unchanged.
  static turbo_stream_fragment(
    action: string,
    target: string,
    html: string,
    attributes: string = "",
  ): string {
    if (action === "remove") {
      return `<turbo-stream${attributes} action="remove" target="${target}"></turbo-stream>`;
    }
    return `<turbo-stream${attributes} action="${action}" target="${target}"><template>${html}</template></turbo-stream>`;
  }

  static prepend(opts: BroadcastOpts): void {
    emit("prepend", opts);
  }
  static append(opts: BroadcastOpts): void {
    emit("append", opts);
  }
  static replace(opts: BroadcastOpts): void {
    emit("replace", opts);
  }
  static remove(opts: BroadcastOpts): void {
    emit("remove", opts);
  }
}

File: runtime/rust/broadcasts.rs
//! Turbo Streams broadcasts shim.
//!
//! The model lowerer's `broadcasts_to` expansion (see
//! `src/lower/broadcasts.rs`) produces calls like
//! `Broadcasts.prepend(stream: "x", target: "y", html: "...")`
//! from inside model callback methods (`after_create`, etc.).
//! rust2 emits the kwargs as a `HashMap<String, Value>`, so each
//! shim method here accepts the unified hash shape and pulls
//! the named fields out.
//!
//! State lives in a thread-local log so framework tests can assert
//! on what got emitted; production installs a broadcaster via
//! `install_broadcaster` that fans fragments out to the cable
//! websocket. Mirrors `runtime/crystal/broadcasts.cr` member-for-
//! member at the shim level.
//!
//! For Phase 5 the production path is a no-op stub — Cable wiring
//! arrives in a later phase.

use std::cell::RefCell;
use std::collections::HashMap;

use serde_json::Value;

thread_local! {
    /// In-memory broadcast log: `(action, stream, target, html)`
    /// tuples in emission order. Tests inspect this after running
    /// model callbacks; production reads it through `log()` if a
    /// fan-out plugin needs the trail.
    static LOG: RefCell<Vec<(String, String, String, String)>> =
        const { RefCell::new(Vec::new()) };
}

/// `Broadcasts` namespace — `pub struct Broadcasts;` + impl gives
/// the same `Broadcasts::method(...)` call shape the lowered model
/// callbacks emit.
pub struct Broadcasts;

impl Broadcasts {
    /// Reset the in-memory log. Framework tests call this between
    /// assertions; production typically doesn't.
    pub fn reset_log_bang() {
        LOG.with(|c| c.borrow_mut().clear());
    }

    /// Snapshot the log as a fresh Vec.
    pub fn log() -> Vec<(String, String, String, String)> {
        LOG.with(|c| c.borrow().clone())
    }

    pub fn append(attrs: HashMap<String, Value>) {
        Self::record("append", &attrs);
    }

    pub fn prepend(attrs: HashMap<String, Value>) {
        Self::record("prepend", &attrs);
    }

    pub fn replace(attrs: HashMap<String, Value>) {
        Self::record("replace", &attrs);
    }

    pub fn remove(attrs: HashMap<String, Value>) {
        Self::record("remove", &attrs);
    }

    fn record(action: &str, attrs: &HashMap<String, Value>) {
        let stream = attrs.get("stream").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let target = attrs.get("target").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let html = attrs.get("html").and_then(|v| v.as_str()).unwrap_or("").to_string();
        LOG.with(|c| {
            c.borrow_mut().push((
                action.to_string(),
                stream.clone(),
                target.clone(),
                html.clone(),
            ))
        });
        // Forward to the live Action Cable fan-out. The model callbacks
        // (`broadcasts_to` lowering) hand us the already-rendered partial
        // as `html`; compose the `<turbo-stream>` wrapper and fan it out to
        // every subscriber of `stream`. Without this the production path is
        // a no-op (only the in-memory LOG is populated) and a subscribed
        // `<turbo-cable-stream-source>` never sees the create/destroy
        // broadcast — the e2e action_cable spec. The cable server fans out
        // via per-subscriber mpsc channels, so this is safe to call from
        // any axum worker thread.
        let fragment = crate::cable::turbo_stream_html(action, &target, &html);
        crate::cable::CABLE.broadcast(&stream, &fragment);
    }
}
